distribution/lib/Standard/Base/0.0.0-dev/src/Meta.enso

```diff
@@ -561,3 +568,24 @@ get_short_type_name typ = @Builtin_Method "Meta.get_short_type_name"
    - constructor: the constructor to get the declaring type of.
 get_constructor_declaring_type : Any -> Any
 get_constructor_declaring_type constructor = @Builtin_Method "Meta.get_constructor_declaring_type"
+
+instrumentor_builtin op args = @Builtin_Method "Meta.instrumentor_builtin"
+
+# Builder to create instrumentation for a function
+type Instrumentor
+    Value impl
+    on_enter self (fn : Text -> Any | Nothing) =
+        instrumentor_builtin "onEnter" [ self.impl, fn ]
+        self
+    on_return self (fn : Text -> Any -> Nothing) =
+        instrumentor_builtin "onReturn" [ self.impl, fn ]
+        self
+    on_function self (fn : Text -> Any | Nothing) =
+        instrumentor_builtin "onFunction" [ self.impl, fn ]
+        self
+    on_exception self (fn : Any -> Nothing) =
+        instrumentor_builtin "onException" [ self.impl, fn ]
+        self
+    activate self =
+        instrumentor_builtin "activate" [ self.impl ]
+        self
```

Comment:
Can we get a documentation on these?

I can't easily tell the difference between `on_function` and `on_enter` for example.

Reply:
These callbacks are inspired by the [methods of IdExecutionService](https://github.com/enso-org/enso/pull/7833/files#diff-1394e5fab22371b78d8206bcbb14124e662b895a7ac9c049f95ee07739111d5aR41). @4e6, do you agree with these names, or do you want to use different, more `IdExecutionService` related ones? How would you describe `on_function` - e.g. `onFunctionReturn` meaning? Is the [Javadoc in onFunctionReturn](https://github.com/enso-org/enso/pull/7833/files#diff-1394e5fab22371b78d8206bcbb14124e662b895a7ac9c049f95ee07739111d5a) OK?

Reply:
So there is a different notification depending on if the returned value is a function or not?

Reply:
The returned JSON-RPC message is always the same but the callbacks have different logic. One difference is that the function one tries to resolve the method pointer of the computed expression.

Reply:
I'm not sure if we want to expose that as an API or not

Reply:
Thank you @4e6, I can hide the `on_function` callback for now. Could you suggest a [modification to the test](https://github.com/enso-org/enso/pull/7833/files#r1342961492) to check `on_function` and/or `on_exception` behavior?

Reply:
`on_exception` should be triggered on panics

Reply:
You can also check that `on_exception` does not catch the `TailCallException` because it is an implementation detail.

Reply:
Turned out I have no idea to write `on_exception` test. Hence I removed the concept in #7970